// vue-codemirror6/dist/index.es.js
/**
 * vue-codemirror6
 *
 * @description CodeMirror6 Component for vue2 and vue3.
 * @author Logue <logue@hotmail.co.jp>
 * @copyright 2022-2025 By Masashi Yoshikawa All rights reserved.
 * @license MIT
 * @version 1.4.1
 * @see {@link https://github.com/logue/vue-codemirror6}
 */

import { indentWithTab as W } from "@codemirror/commands";
import { indentUnit as G } from "@codemirror/language";
import { linter as P, lintGutter as Z, forceLinting as w, diagnosticCount as q } from "@codemirror/lint";
import { EditorState as u, Compartment as C, StateEffect as S, EditorSelection as O } from "@codemirror/state";
import { EditorView as d, placeholder as H, keymap as K } from "@codemirror/view";
import { basicSetup as Q, minimalSetup as X } from "codemirror";
import { isVue2 as Y, h as V, defineComponent as ee, ref as m, shallowRef as te, computed as f, watch as B, onMounted as ae, nextTick as ne, onUnmounted as ie } from "vue-demi";
const he = {
  version: "1.4.1",
  date: "2025-10-17T22:30:13.688Z"
}, le = (t) => t ? Object.entries(t).reduce((l, [i, s]) => (i = i.charAt(0).toUpperCase() + i.slice(1), i = `on${i}`, { ...l, [i]: s }), {}) : {};
function M(t, l = {}, i) {
  if (Y)
    return V(t, l, i);
  const { props: s, domProps: e, on: v, ...r } = l, h = v ? le(v) : {};
  return V(
    t,
    { ...r, ...s, ...e, ...h },
    i
  );
}
const oe = (t) => typeof t == "function" ? t() : t, ue = ee({
  /** Component Name */
  name: "CodeMirror",
  /** Model Definition */
  model: {
    prop: "modelValue",
    event: "update:modelValue"
  },
  /** Props Definition */
  props: {
    /** Model value */
    modelValue: {
      type: String,
      default: ""
    },
    /**
     * Theme
     *
     * @see {@link https://codemirror.net/docs/ref/#view.EditorView^theme}
     */
    theme: {
      type: Object,
      default: () => ({})
    },
    /** Dark Mode */
    dark: {
      type: Boolean,
      default: !1
    },
    /**
     * Use Basic Setup
     *
     * This will enable the basic setup for the editor.
     *
     * @see {@link https://codemirror.net/docs/ref/#codemirror.basicSetup}
     */
    basic: {
      type: Boolean,
      default: !1
    },
    /**
     * Use Minimal Setup (The basic setting has priority.)
     *
     * @see {@link https://codemirror.net/docs/ref/#codemirror.minimalSetup}
     */
    minimal: {
      type: Boolean,
      default: !1
    },
    /**
     * Placeholder
     *
     * @see {@link https://codemirror.net/docs/ref/#view.placeholder}
     */
    placeholder: {
      type: String,
      default: void 0
    },
    /**
     * Line wrapping
     *
     * An extension that enables line wrapping in the editor (by setting CSS white-space to pre-wrap in the content).
     *
     * @see {@link https://codemirror.net/docs/ref/#view.EditorView%5ElineWrapping}
     */
    wrap: {
      type: Boolean,
      default: !1
    },
    /**
     * Allow tab key indent.
     *
     * This will enable the tab key to indent the current line or selection.
     *
     * @see {@link https://codemirror.net/examples/tab/}
     */
    tab: {
      type: Boolean,
      default: !1
    },
    /**
     * Tab character
     *
     * This is the unit of indentation used when the editor is configured to indent with tabs.
     * It is also used to determine the size of the tab character when the editor is configured to use tabs for indentation..
     *
     * @see {@link https://codemirror.net/docs/ref/#state.EditorState^indentUnit}
     */
    indentUnit: {
      type: String,
      default: void 0
    },
    /**
     * Allow Multiple Selection.
     *
     * This allows the editor to have multiple selections at the same time.
     * This is useful for editing multiple parts of the document at once.
     * If this is set to true, the editor will allow multiple selections.
     * If this is set to false, the editor will only allow a single selection.
     *
     * @see {@link https://codemirror.net/docs/ref/#state.EditorState^allowMultipleSelections}
     */
    allowMultipleSelections: {
      type: Boolean,
      default: !1
    },
    /**
     * Tab size
     *
     * This is the number of spaces that a tab character represents in the editor.
     * It is used to determine the size of the tab character when the editor is configured to use tabs for indentation.
     * If this is set to a number, the editor will use that number of spaces for each tab character.
     *
     * @see {@link https://codemirror.net/docs/ref/#state.EditorState^tabSize}
     */
    tabSize: {
      type: Number,
      default: void 0
    },
    /**
     * Set line break (separetor) char.
     *
     * This is the character that is used to separate lines in the editor.
     * It is used to determine the line break character when the editor is configured to use a specific line break character.
     *
     * @see {@link https://codemirror.net/docs/ref/#state.EditorState^lineSeparator}
     */
    lineSeparator: {
      type: String,
      default: void 0
    },
    /**
     * Readonly
     *
     * This is a CodeMirror Facet that allows you to set the editor to read-only mode.
     * When this is set to true, the editor will not allow any changes to be made to the document.
     * This is useful for displaying code that should not be edited, such as documentation or examples.
     * If this is set to false, the editor will allow changes to be made to the document.
     *
     * @see {@link https://codemirror.net/docs/ref/#state.EditorState^readOnly}
     */
    readonly: {
      type: Boolean,
      default: !1
    },
    /**
     * Disable input.
     *
     * This is the reversed value of the CodeMirror editable.
     * Similar to `readonly`, but setting this value to true disables dragging.
     *
     * @see {@link https://codemirror.net/docs/ref/#view.EditorView^editable}
     */
    disabled: {
      type: Boolean,
      default: !1
    },
    /**
     * Additional Extension
     *
     * You can use this to add any additional extensions that you want to use in the editor.
     *
     * @see {@link https://codemirror.net/docs/ref/#state.Extension}
     */
    extensions: {
      type: Array,
      default: () => []
    },
    /**
     * Language Phreses
     *
     * This is a CodeMirror Facet that allows you to define custom phrases for the editor.
     * It can be used to override default phrases or add new ones.
     * This is useful for translating the editor to different languages or for customizing the editor's UI.
     *
     * @see {@link https://codemirror.net/examples/translate/}
     */
    phrases: {
      type: Object,
      default: void 0
    },
    /**
     * CodeMirror Language
     *
     * This is a CodeMirror Facet that allows you to define the language of the editor.
     * It can be used to enable syntax highlighting and other language-specific features.
     * It is useful for displaying code in a specific language, such as JavaScript, Python, or HTML.
     *
     * @see {@link https://codemirror.net/docs/ref/#language}
     */
    lang: {
      type: Object,
      default: void 0
    },
    /**
     * CodeMirror Linter
     *
     * This is a CodeMirror Facet that allows you to define a linter for the editor.
     * It can be used to check the code for errors and warnings, and to provide feedback to the user.
     * It is useful for displaying code in a specific language, such as JavaScript, Python, or HTML.
     * This is useful for providing feedback to the user about the code they are writing.
     *
     * @see {@link https://codemirror.net/docs/ref/#lint.linter}
     */
    linter: {
      type: Function,
      default: void 0
    },
    /**
     * Linter Config
     *
     * This is a CodeMirror Facet that allows you to define the configuration for the linter.
     * It can be used to specify options for the linter, such as the severity of errors and warnings, and to customize the behavior of the linter.
     * This is useful for providing feedback to the user about the code they are writing.
     *
     * @see {@link https://codemirror.net/docs/ref/#lint.linter^config}
     */
    linterConfig: {
      type: Object,
      default: () => ({})
    },
    /**
     * Forces any linters configured to run when the editor is idle to run right away.
     *
     * This is useful for running linters on the initial load of the editor, or when the user has made changes to the code and wants to see the results immediately.
     *
     * @see {@link https://codemirror.net/docs/ref/#lint.forceLinting}
     */
    forceLinting: {
      type: Boolean,
      default: !1
    },
    /**
     * Show Linter Gutter
     *
     * An area to 🔴 the lines with errors will be displayed.
     * This feature is not enabled if `linter` is not specified.
     *
     * @see {@link https://codemirror.net/docs/ref/#lint.lintGutter}
     */
    gutter: {
      type: Boolean,
      default: !1
    },
    /**
     * Gutter Config
     *
     * This is a CodeMirror Facet that allows you to define the configuration for the gutter.
     * It can be used to specify options for the gutter, such as the size of the gutter, the position of the gutter, and to customize the behavior of the gutter.
     * This is useful for providing feedback to the user about the code they are writing.
     *
     * @see {@link https://codemirror.net/docs/ref/#lint.lintGutter^config}
     */
    gutterConfig: {
      type: Object,
      default: void 0
    },
    /**
     * Using tag
     */
    tag: {
      type: String,
      default: "div"
    },
    /**
     * Allows an external update to scroll the form.
     *
     * This is useful for scrolling the editor to a specific position when the user has made changes to the code and wants to see the results immediately.
     * If this is set to true, the editor will scroll to the position specified in the transaction.
     * If this is set to false, the editor will not scroll to the position specified in the transaction.
     *
     * @see {@link https://codemirror.net/docs/ref/#state.TransactionSpec.scrollIntoView}
     */
    scrollIntoView: {
      type: Boolean,
      default: !0
    },
    /**
     * Key map
     * This is a CodeMirror Facet that allows you to define custom key bindings.
     * It can be used to override default key bindings or add new ones.
     *
     * @see {@link https://codemirror.net/docs/ref/#view.keymap}
     */
    keymap: {
      type: Array,
      default: () => []
    }
  },
  /** Emits */
  emits: {
    /** Model Update */
    "update:modelValue": (t = "") => !0,
    /** CodeMirror ViewUpdate */
    update: (t) => !0,
    /** CodeMirror onReady */
    ready: (t) => !0,
    /** CodeMirror onFocus */
    focus: (t) => !0,
    /** State Changed */
    change: (t) => !0,
    /** CodeMirror onDestroy */
    destroy: () => !0
  },
  /**
   * Setup
   *
   * @param props  - Props
   * @param context - Context
   */
  setup(t, l) {
    const i = m(), s = m(t.modelValue), e = te(void 0), v = f({
      get: () => e.value?.hasFocus ?? !1,
      set: (a) => {
        a && e.value && e.value.focus();
      }
    }), r = f({
      get: () => e.value?.state.selection,
      set: (a) => {
        e.value && a && e.value.dispatch({ selection: a });
      }
    }), h = f({
      get: () => e.value?.state.selection.main.head ?? 0,
      set: (a) => {
        e.value && e.value.dispatch({ selection: { anchor: a } });
      }
    }), j = f({
      get: () => e.value?.state.toJSON(),
      set: (a) => {
        e.value && a && e.value.setState(u.fromJSON(a));
      }
    }), p = m(0), g = m(0), y = f(() => {
      const a = new C(), n = new C();
      if (t.basic && t.minimal)
        throw new Error(
          "[Vue CodeMirror] Both basic and minimal cannot be specified."
        );
      let o = [];
      return t.keymap && t.keymap.length > 0 && (o = t.keymap), t.tab && o.push(W), [
        // Toggle basic setup
        t.basic && !t.minimal ? Q : void 0,
        // Toggle minimal setup
        t.minimal && !t.basic ? X : void 0,
        // ViewUpdate event listener
        d.updateListener.of((c) => {
          e.value && (l.emit("focus", e.value.hasFocus), p.value = e.value.state.doc?.length, !(c.changes.empty || !c.docChanged) && (t.linter && (t.forceLinting && w(e.value), g.value = t.linter(e.value).length), l.emit("update", c)));
        }),
        // Toggle light/dark mode.
        d.theme(t.theme, { dark: t.dark }),
        // Toggle line wrapping
        t.wrap ? d.lineWrapping : void 0,
        // Tab character
        t.indentUnit ? G.of(t.indentUnit) : void 0,
        // Allow Multiple Selections
        u.allowMultipleSelections.of(t.allowMultipleSelections),
        // Indent tab size
        t.tabSize ? n.of(u.tabSize.of(t.tabSize)) : void 0,
        // locale settings
        t.phrases ? u.phrases.of(t.phrases) : void 0,
        // Readonly option
        u.readOnly.of(t.readonly),
        // Editable option
        d.editable.of(!t.disabled),
        // Set Line break char
        t.lineSeparator ? u.lineSeparator.of(t.lineSeparator) : void 0,
        // Lang
        t.lang ? a.of(t.lang) : void 0,
        // Append Linter settings
        t.linter ? P(t.linter, t.linterConfig) : void 0,
        // Show 🔴 to error line when linter enabled.
        t.linter && t.gutter ? Z(t.gutterConfig) : void 0,
        // Placeholder
        t.placeholder ? H(t.placeholder) : void 0,
        // Keymap and Indent with Tab
        o.length > 0 ? K.of(o) : void 0,
        // Append Extensions
        ...t.extensions
      ].filter((c) => !!c);
    });
    B(
      y,
      (a) => e.value?.dispatch({ effects: S.reconfigure.of(a) }),
      { immediate: !0 }
    ), B(
      () => t.modelValue,
      async (a) => {
        if (!e.value || e.value.composing || // IME fix
        e.value.state.doc.toJSON().join(t.lineSeparator ?? `
`) === a)
          return;
        const n = !e.value.state.selection.ranges.every(
          (o) => o.anchor < a.length && o.head < a.length
        );
        e.value.dispatch({
          changes: { from: 0, to: e.value.state.doc.length, insert: a },
          selection: n ? { anchor: 0, head: 0 } : e.value.state.selection,
          scrollIntoView: t.scrollIntoView
        });
      },
      { immediate: !0 }
    ), ae(async () => {
      if (globalThis.window === void 0 || !i.value)
        return;
      let a = s.value;
      i.value.childNodes[0] && (s.value !== "" && console.warn(
        "[CodeMirror.vue] The <code-mirror> tag contains child elements that overwrite the `v-model` values."
      ), a = i.value.childNodes[0].innerText.trim()), e.value = new d({
        parent: i.value,
        state: u.create({ doc: a, extensions: y.value }),
        dispatch: (n) => {
          e.value && (e.value.update([n]), !(n.changes.empty || !n.docChanged) && (l.emit("update:modelValue", n.state.doc.toString()), l.emit("change", n.state)));
        }
      }), await ne(), l.emit("ready", {
        view: e.value,
        state: e.value.state,
        container: i.value
      });
    }), ie(() => {
      e.value && (e.value.destroy(), l.emit("destroy"));
    });
    const x = () => {
      !t.linter || !e.value || (t.forceLinting && w(e.value), g.value = q(e.value.state));
    }, L = () => {
      e.value?.dispatch({
        effects: S.reconfigure.of([])
      }), e.value?.dispatch({
        effects: S.appendConfig.of(y.value)
      });
    }, N = (a, n) => e.value?.state.sliceDoc(a, n), T = (a) => e.value?.state.doc.line(a + 1).text, U = () => e.value?.state.doc.lines ?? 0, $ = () => e.value?.state.selection.main.head ?? 0, k = () => e.value?.state.selection.ranges ?? [], z = () => e.value ? e.value.state.sliceDoc(
      e.value.state.selection.main.from,
      e.value.state.selection.main.to
    ) : "", E = () => {
      const a = e.value?.state;
      return a ? a.selection.ranges.map(
        (n) => a.sliceDoc(n.from, n.to)
      ) : [];
    }, R = () => e.value?.state.selection.ranges.some(
      (a) => !a.empty
    ) ?? !1, _ = (a, n, o) => {
      e.value && e.value.dispatch({
        changes: { from: n, to: o, insert: a }
      });
    }, D = (a) => {
      e.value && e.value.dispatch(e.value.state.replaceSelection(a));
    }, A = (a) => {
      e.value && e.value.dispatch({ selection: { anchor: a } });
    }, F = (a, n) => {
      e.value && e.value.dispatch({ selection: { anchor: a, head: n } });
    }, I = (a, n) => {
      e.value && e.value.dispatch({
        selection: O.create(a, n)
      });
    }, J = (a) => {
      e.value && r.value && e.value.dispatch({
        selection: O.create(
          r.value.ranges.map((n) => n.extend(a(n)))
        )
      });
    }, b = {
      editor: i,
      view: e,
      cursor: h,
      selection: r,
      focus: v,
      length: p,
      json: j,
      diagnosticCount: g,
      dom: e.value?.contentDOM,
      lint: x,
      forceReconfigure: L,
      // Bellow is CodeMirror5's function
      getRange: N,
      getLine: T,
      lineCount: U,
      getCursor: $,
      listSelections: k,
      getSelection: z,
      getSelections: E,
      somethingSelected: R,
      replaceRange: _,
      replaceSelection: D,
      setCursor: A,
      setSelection: F,
      setSelections: I,
      extendSelectionsBy: J
    };
    return l.expose(b), b;
  },
  render() {
    return M(
      this.$props.tag,
      {
        ref: "editor",
        class: "vue-codemirror"
      },
      this.$slots.default ? (
        // Hide original content
        M(
          "aside",
          { style: "display: none;", "aria-hidden": "true" },
          oe(this.$slots.default)
        )
      ) : void 0
    );
  }
}), ge = (t) => t.component("CodeMirror", ue);
export {
  he as Meta,
  ue as default,
  ge as install
};
